### PR TITLE
Introducing WhitespaceBeforeArrayInitializerCheck, fixing issue #333

### PR DESCRIPTION
**Pull request by liscju**
Introduces new WhitespaceBeforeArrayInitializerCheck. When we deal with Array Initializer we have to deal with two situations:
a) int ints[] = {1, 2, 3, 4};
b) new int[] {1,2,3,4}
In first situation we have to check where lays the parent of Array Initializer(Assignment) , in second we have to check previous sibling (Array Declarator int[]).

### CODE REVIEW

**Comment by romani** [comment]
I beg apology, I extremely occupied at main project (even with PRs from you).  I will do my best to reply you before Monday.

File: sevntu-checks/pom.xml [hunk 1]
```diff
@@ -113,6 +113,7 @@
                     <regex><pattern>.*.checks.coding.ReturnNullInsteadOfBoolean</pattern><branchRate>80</branchRate><lineRate>88</lineRate></regex>
                     <regex><pattern>.*.checks.coding.SimpleAccessorNameNotationCheck</pattern><branchRate>75</branchRate><lineRate>100</lineRate></regex>
                     <regex><pattern>.*.checks.coding.UnnecessaryParenthesesExtendedCheck</pattern><branchRate>93</branchRate><lineRate>100</lineRate></regex>
+                    <regex><pattern>.*.checks.coding.WhitespaceBeforeArrayInitializerCheck</pattern><branchRate>87</branchRate><lineRate>100</lineRate></regex>
```
Comment (mkordas) [inline sevntu-checks/pom.xml]: Let's bring coverage to 100%, this should be the rule for all new checks

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/WhitespaceBeforeArrayInitializerCheck.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * <p>
+ * This checks enforces whitespace before array initializer
+ * </p>
+ * Examples:
+ * This code is perfectly valid:
+ * <p>
+ * <code>
+ * <pre>
+ * int[] ints = new int[] {
+ *           0, 1, 2, 3
```
Comment (mkordas) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/WhitespaceBeforeArrayInitializerCheck.java]: I wouldn't say it's perfectly valid, it has 10 space indentation ;)

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/WhitespaceBeforeArrayInitializerCheck.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+/**
+ * <p>
+ * This checks enforces whitespace before array initializer
+ * </p>
+ * Examples:
+ * This code is perfectly valid:
+ * <p>
+ * <code>
+ * <pre>
+ * int[] ints = new int[] {
```
Comment (mkordas) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/WhitespaceBeforeArrayInitializerCheck.java]: we also need to state that 
```java
new int[]
    {0, 1, 2, 3}
```
is also correct

**Comment by mkordas** [comment]
Please validate sevntu codebase using this check and fix all violations in separate commit. Please generate report e.g. for Guava.

File: sevntu-checks/src/main/resources/com/github/sevntu/checkstyle/checks/coding/messages.properties [hunk 4]
```diff
@@ -106,3 +106,4 @@ useless.super.ctor.call.in.not.derived.class= Super call could be removed: Class
 useless.super.ctor.call.without.args= Redundant super constructor call could be removed. Class ''{0}'' has superclass. Java compiler automatically inserts a call to the no-argument constructor of the superclass.
 return.null.Boolean=Method declares to return Boolean and returns null.
 return.boolean.ternary=Returning explicit boolean from ternary operator.
+whitespace.before.array.initializer=Array initializer should have whitespace before
```
Comment (mkordas) [inline sevntu-checks/src/main/resources/com/github/sevntu/checkstyle/checks/coding/messages.properties]: no dot at the end like other messages have

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/WhitespaceBeforeArrayInitializerCheck.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        return new int[] {TokenTypes.ARRAY_INIT};
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        if (ast.getPreviousSibling() != null && ast.getPreviousSibling().getType() == TokenTypes.ARRAY_DECLARATOR) {
+            DetailAST arrayDeclaratorAST = ast.getPreviousSibling();
+            checkDistanceBetweenTokens(ast, getEndOfArrayDeclarator(arrayDeclaratorAST));
+        } else {
+            //assert ast.getParent().getType() == TokenTypes.ASSIGN;
```
Comment (mkordas) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/WhitespaceBeforeArrayInitializerCheck.java]: Do not commit commented out code, that should be the rule

Comment (romani) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/WhitespaceBeforeArrayInitializerCheck.java]: @mkordas , agree with you but do not know how to make algorithm that distinguish commented code and just a comment with some code in it.

File: sevntu-checks/pom.xml (hunk 1 above)
Comment (romani) [inline sevntu-checks/pom.xml]: new Check should appear with 100% coverage , in other words no update is required in pom.xml. That lines are here only for old Checks that existed before we enforced 100%.

**Comment by romani** [comment]
not all files are updated, please review commit https://github.com/sevntu-checkstyle/sevntu.checkstyle/commit/3f13341b98e1b36d9cb65111dbb1bd7bbb2206c7 for all files that should be changed. Sorry that such process is not automated yet.

I will give you detailed code review in 24 hours.

**Comment by romani** [comment]
0) do all notes from mkordas

1) please do Javadoc as main project is required, that Check should eventually appear in main project


2) 
> ```
if (ast.getPreviousSibling() != null && ast.getPreviousSibling().getType() 
```

please spit into two lines. and do "ast.getPreviousSibling()" ones.

3)
> private void checkDistanceBetweenTokens(

Plase explain me requrement of "abs" and please be lazy in calculations, calculate column first and comapre it and then lines.

4) 
> firstAST.getColumnNo()-secondAST.getColumnNo(

please use spaces between "-"

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/WhitespaceBeforeArrayInitializerCheck.java (hunk 5 above)
Comment (mkordas) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/WhitespaceBeforeArrayInitializerCheck.java]: @romani, if someone really wants to put some code as documentation, then Javadoc format can be used

Comment (mkordas) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/WhitespaceBeforeArrayInitializerCheck.java]: @romani there is blog post about it done by Sonar: http://www.sonarqube.org/commented-out-code-eradication-with-sonar/

Comment (romani) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/WhitespaceBeforeArrayInitializerCheck.java]: Ok , put an issues to reuse sonar experience at checkstyle :) .

**Comment by romani** [comment]
@liscju , do you plan to finish that issue ?

**Comment by liscju** [comment]
@romani yes, i was really busy last three weeks so my PRs for checkstyle and sevntu.checkstyle stay unfinished. Im gonna work on it in weekend

**Comment by romani** [comment]
@liscju , do you still plan to finish that issue ?
It is ok to drop PR, I just want to let you finish before redirecting that PR to other contributor.

**Comment by romani** [comment]
@liscju , do you still plan to finish that issue ?

I am planning to do 1.14 in next week or two.

**Comment by romani** [comment]
@liscju , do you still plan to finish that issue ?

**Comment by liscju** [comment]
@romani :
0) all @mkordas suggestions applied , guava reports is here: http://liscju.github.io/WhitespaceBeforeArrayInitializer/target/site/checkstyle.html and all violations is "valid"

1) i dont understand, can you provide more explanation?

2) not applicable in this version

3) now abs is not used, lazy evaluation done as requested

4) DONE

**Comment by romani** [comment]
@liscju ,

1) 
> ```<regex><pattern>.*.checks.coding.WhitespaceBeforeArrayInitializerCheck</pattern><branchRate>100</branchRate><lineRate>100</lineRate></regex>``

just remove this line , that configuration is required only for checks that are not covered as 100%


2) 
> private void checkDistanceBetweenTokens(DetailAST firstAST, DetailAST secondAST) 

please refactor this method to be static and return boolean value. "log"  should be done in visitToken. By mean of this you will separate algorithm and chekstyle's stuff.

3) Report is good, please apply minor changes from above and I am ready to merge your code.

**Comment by liscju** [comment]
1) DONE , deleted WhiteSpaceBeforeArrayInitializer coverage requirement from pom.xml

2) DONE, made static returning boolean , log moved to visitToken

**Comment by romani** [comment]
1)
> private static boolean isDistanceBetweenAstValid(DetailAST firstAST, DetailAST secondAST)

Valid ? please avoid such general words as much as possible. Please be exact - areTokensSeparatedByWhitespace(token, token).  By means of this algorithm will be written by almost-human language in visitToken and method is named exactly as what it do.

**Comment by liscju** [comment]
@romani Right, renamed it as requested

**Comment by romani** [comment]
thanks a LOT ! I will release you during weekends.

**Pull request merged**